components/builder-core/src/metrics.rs

```diff
@@ -24,44 +25,57 @@ pub const APP_NAME: &'static str = "bldr";
 // Statsd Listener Address
 pub const STATS_ENV: &'static str = "HAB_STATS_ADDR";
 
-// Supported metrics
-#[derive(Debug, Clone)]
-pub enum Counter {
-    GithubApi,
-    GithubAuthenticate,
-    GithubEvent,
-    GithubInstallationToken,
-    SearchPackages,
+pub type ApiEndpoint = &'static str;
+pub type InstallationId = u32;
+
+// Public Interface
+////////////////////////////////////////////////////////////////////////
+
+/// Metric identifiers will usually be static `str`s, but some may
+/// need to be dynamically-generated `String`s. With a `Cow`, we can
+/// accept either.
+pub type MetricId = Cow<'static, str>;
+
+/// All metrics must implement the Metric trait, as well as one of the
+/// type marker traits (e.g., `CounterMetric`).
+pub trait Metric {
+    /// Generate the metric name to be used
+    fn id(&self) -> MetricId;
 }
 
-// Supported metrics
-#[derive(Debug, Clone)]
-pub enum Gauge {
-    PackageCount,
+/// Marker trait for counter metrics (don't want to accidentally
+/// increment a gauge metric!)
+pub trait CounterMetric {}
+
+/// Increment the given metric by one
+pub fn incr<C>(counter: C)
```

Comment:
Should incr be defined as a CounterMetric trait fn?

Reply:
Possibly? I didn't do it here to ensure that all the implementation was concentrated here. The idea is that you declare minimal types for your own metrics, and then pass those as input to one core set of functions. 

We could probably achieve the same thing with default implementations for `incr` and `decr` in `CounterMetric`.  I kinda wanted to see how this way played out for a little while. Also, I found a [dogstatsd](https://github.com/mcasper/dogstatsd-rs) crate that I was thinking about trying out next to get tagging, so I will probably be back in here soon anyway.

Reply:
Right, the minimal type is goodness.. my thinking though is, if we defined incr as a default fn on the CounterMetric trait, then you'd never need to know about the global incr function at all... you'd just declare your Counter enum, impl the id function, and just call incr on it.